### PR TITLE
Mark gcc-<version> as `gcc` instead of `compiler` in Unix CC toolchain

### LINKED ISSUE #17794: @bazel_tools//tools/cpp:compiler:gcc detection does not handle `gcc-10`
### Description of the bug:

Bazel's detection of GCC only works if the compiler executable name is exactly "gcc":

https://github.com/bazelbuild/bazel/blob/f06d00b8060b3b50f5dbc4a03b957c269ef62670/tools/cpp/unix_cc_configure.bzl#L273

Consequently, it fails to detect common variations like `gcc-10` as GCC.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

CC=gcc-10 bazel ...

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 6.1.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by robinlinden**
Fixes #17794

I was looking at writing a test for this, but not sure how you'd like me to go about that. Hard-code a gcc, e.g. `gcc-11`, or search the `$PATH` for a `gcc-<version>`?

### CODE REVIEW

File: tools/cpp/unix_cc_configure.bzl [hunk 1]
```diff
@@ -273,7 +273,14 @@ def _is_clang(repository_ctx, cc):
 def _is_gcc(repository_ctx, cc):
     # GCC's version output uses the basename of argv[0] as the program name:
     # https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/gcc.cc;h=158461167951c1b9540322fb19be6a89d6da07fc;hb=HEAD#l8728
-    return repository_ctx.execute([cc, "--version"]).stdout.startswith("gcc ")
+    cc_argv0 = repository_ctx.execute([cc, "--version"]).stdout.partition(" ")[0]
+    if (cc_argv0 == "gcc"):
```
Comment (fmeum) [inline tools/cpp/unix_cc_configure.bzl:277]:
```suggestion
    if cc_argv0 == "gcc":
```

File: tools/cpp/unix_cc_configure.bzl [hunk 2]
```diff
@@ -273,7 +273,14 @@ def _is_clang(repository_ctx, cc):
...
+        return True
+
+    if (cc_argv0.startswith("gcc-")):
```
Comment (fmeum) [inline tools/cpp/unix_cc_configure.bzl:280]:
```suggestion
    if cc_argv0.startswith("gcc-"):
```

File: tools/cpp/unix_cc_configure.bzl [hunk 3]
```diff
@@ -273,7 +273,14 @@ def _is_clang(repository_ctx, cc):
...
+        return cc_argv0.removeprefix("gcc-").isdigit()
```
Comment (fmeum) [inline tools/cpp/unix_cc_configure.bzl:281]: I think that starting with `gcc-` is probably good enough of an indicator without a separate check of the characters that follow. In fact, an executable such as `gcc-arm-none-eabi` should ideally also be recognized as `gcc`.

**Review by fmeum — COMMENTED** [review]
Thanks for taking this on! Just a few drive-by comments, this will have to be reviewed by a team member.

You can run `bazel run //src/test/tools/bzlmod:update_default_lock_file` and commit the result to address the CI failures.

**Comment by robinlinden** [comment]
> Thanks for taking this on! Just a few drive-by comments, this will have to be reviewed by a team member.
> 
> You can run `bazel run //src/test/tools/bzlmod:update_default_lock_file` and commit the result to address the CI failures.

Thank you for the initial gcc-marking, as well as having a look and explaining the lock-file updating! I had a look at a few random passing PRs and didn't see any lock-file updates, so I figured I just had the worst of luck. The drive-by comments are very helpful. :)

File: tools/cpp/unix_cc_configure.bzl (hunk 3 above)
Comment (robinlinden) [reply tools/cpp/unix_cc_configure.bzl]: Dropped the `isdigit()`-check!

File: tools/cpp/unix_cc_configure.bzl [hunk 4]
```diff
@@ -273,7 +273,15 @@ def _is_clang(repository_ctx, cc):
 def _is_gcc(repository_ctx, cc):
     # GCC's version output uses the basename of argv[0] as the program name:
     # https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/gcc.cc;h=158461167951c1b9540322fb19be6a89d6da07fc;hb=HEAD#l8728
-    return repository_ctx.execute([cc, "--version"]).stdout.startswith("gcc ")
+    cc_argv0 = repository_ctx.execute([cc, "--version"]).stdout.partition(" ")[0]
+    if cc_argv0 == "gcc":
```
Comment (comius) [inline tools/cpp/unix_cc_configure.bzl:277]: Please change to `x.startswith("gcc ") or x.startswith("gcc-")` without using call to `partition` first.

Check without using `partition` should be more robust. One thing less that can go wrong. (But I'm also nitpicking a bit).

File: MODULE.bazel.lock [hunk 5]
```diff
@@ -2248,7 +2248,7 @@
       "general": {
         "bzlTransitiveDigest": "/h6B2Wl9PVY+Bv0PAxzkZUmm947eWCyYthDYT5CPX+k=",
         "accumulatedFileDigests": {
-          "@@//src/test/tools/bzlmod:MODULE.bazel.lock": "e7caa4d6182abd31642ac06dfc30a4501a998924fbd68b59eced1e1cf836da5f",
+          "@@//src/test/tools/bzlmod:MODULE.bazel.lock": "12e9ba281d672e202359d8f7f740b30fca190012aaa24a7ac04c547233534e74",
```
Comment (comius) [inline MODULE.bazel.lock:2251]: I don't know what's current best practice, but I'm guessing MODULE.bazel.lock should not get merged with arbitrary PR. Please remove it.

**Review by comius — CHANGES_REQUESTED** [review]

File: MODULE.bazel.lock (hunk 5 above)
Comment (fmeum) [reply MODULE.bazel.lock:2251]: @comius This is probably necessary: Changing `unix_cc_configure.bzl` changes the transitive bzl digest of the `local_config_cc` module extension used in `bazel_tools`, which in turn causes the `MODULE.bazel.lock` for the tools module (used in tests) to change, which in turn causes the root module lockfile as the `bazel_build_deps` extension parses this lockfile.

Comment (comius) [reply MODULE.bazel.lock:2251]: I thought I wasn't up to date. Thanks for clarification.

File: tools/cpp/unix_cc_configure.bzl (hunk 4 above)
Comment (robinlinden) [reply tools/cpp/unix_cc_configure.bzl]: Thanks for having a look! Rebased and updated. :)

**Review by comius — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.1.0

**Pull request closed without merging**